### PR TITLE
Vault Query Aggregate Function support

### PR DESCRIPTION
**Pull request by josecoll**
Extended VaultCustomQueryCriteria and associated DSL to enable specification of Aggregate Functions (sum, max, min, avg, count) with, optional, group by clauses and sorting (on calculated aggregate)

### CODE REVIEW

**Comment by mikehearn** [comment]
Needs documentation (both api-level and in the docsite). But I'm kind of surprised we're going this route .... isn't this leading us down the reinvent-SQL path? I thought the vault query API was only really meant for the most common kinds of queries and direct SQL could do the rest?

Also, maybe whack a file name annotation on QueryCriteria to avoid the name `QueryCriteriaKt` leaking into Java code (the fact that it's written in Kotlin shouldn't be visible to API users).

**Comment by josecoll** [comment]
JVM filename annotation fails:
adding `@file:JvmName("QueryCriteria")`
causes compilation errors as can no longer reference the `infix` functions `and` `or`

**Comment by josecoll** [comment]
Aggregate Functions have been requested as a requirement by @mnesbit, and also to be used as a replacement for our Cash Balances table (eg. no longer persist cash balances and update per cash transaction update)

File: core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteriaUtils.kt [hunk 1]
```diff
@@ -127,7 +143,7 @@ data class Sort(val columns: Collection<SortColumn>) {
     @CordaSerializable
     enum class Direction {
         ASC,
-        DESC
+        DESC,
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteriaUtils.kt]: Why is there a comma here?

File: node/src/main/kotlin/net/corda/node/services/vault/HibernateQueryCriteriaParser.kt [hunk 2]
```diff
@@ -258,7 +306,14 @@ class HibernateQueryCriteriaParser(val contractType: Class<out ContractState>,
             val joinPredicate = criteriaBuilder.equal(vaultStates.get<PersistentStateRef>("stateRef"), entityRoot.get<PersistentStateRef>("stateRef"))
             joinPredicates.add(joinPredicate)
 
-            predicateSet.add(expressionToPredicate(entityRoot, criteria.expression))
+            // resolve general criteria expressions
+            try {
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/vault/HibernateQueryCriteriaParser.kt]: What's with this try/catch ?

Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/vault/HibernateQueryCriteriaParser.kt]: With regards to our discussions, I don't think the if-then-else approach is very elegant due to heavy usage of generics:

```
            if (criteria.expression is CriteriaExpression.AggregateFunctionExpression<*,*>)
                parseAggregateFunction(entityRoot, criteria.expression as CriteriaExpression.AggregateFunctionExpression<L, Any>)
            else
                predicateSet.add(parseExpression(entityRoot, criteria.expression) as Predicate)
```
I've streamlined the expression parsing code - now called `parseExpression` to deal with all `CriteriaExpression` types. With the exception of Aggregate Functions (which produce select expressions), the parser always returns `Predicate`, so we can replace the try-catch code with:

```
            val expression = parseExpression(entityRoot, criteria.expression)
            if (expression is Predicate)
                predicateSet.add(expression)

```

File: node/src/main/kotlin/net/corda/node/services/vault/HibernateQueryCriteriaParser.kt [hunk 3]
```diff
@@ -254,7 +292,10 @@ class HibernateQueryCriteriaParser(val contractType: Class<out ContractState>,
             val joinPredicate = criteriaBuilder.equal(vaultStates.get<PersistentStateRef>("stateRef"), entityRoot.get<PersistentStateRef>("stateRef"))
             joinPredicates.add(joinPredicate)
 
-            predicateSet.add(expressionToPredicate(entityRoot, criteria.expression))
+            // resolve general criteria expressions
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/vault/HibernateQueryCriteriaParser.kt]: I was thinking replace with a line like this

`parseExpression(entityRoot, criteria.expression, predicateSet)`

and then add a method 

`fun parseExpression(....) {`
`if(criteria.expression is CriteriaExpression.AggregateFunctionExpression<O,*>) {`
`    parseAggregateFunction(root, expression)`
`} else {`
`    predicateSet.add(parseExpression(root, expression) as Predicate)`
`}`

than the case for `CriteriaExpression.AggregateFunctionExpression`  in the `when` can just throw an exception

Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/vault/HibernateQueryCriteriaParser.kt]: Done

**Pull request merged**